File: packages/nhost-js/src/session/__tests__/session.test.ts
import { describe, expect, it } from '@jest/globals';
import { decodeUserSession } from '../session';

/**
 * Encodes a string to base64url (RFC 4648 Section 5), using only vanilla JS APIs.
 */
function toBase64Url(str: string): string {
  const bytes = new TextEncoder().encode(str);
  const binaryString = Array.from(bytes, (b) => String.fromCharCode(b)).join(
    '',
  );
  return btoa(binaryString)
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
}

/**
 * Creates a fake JWT token with the given payload, using base64url encoding
 * (as per the JWT specification, RFC 7519 / RFC 4648 Section 5).
 */
function createJWT(payload: Record<string, unknown>): string {
  const header = toBase64Url(JSON.stringify({ alg: 'HS256', typ: 'JWT' }));
  const body = toBase64Url(JSON.stringify(payload));
  const signature = toBase64Url('fake-signature');
  return `${header}.${body}.${signature}`;
}

describe('decodeUserSession', () => {
  it('should decode a standard JWT token', () => {
    const payload = {
      sub: 'user-123',
      iat: 1700000000,
      exp: 1700003600,
      iss: 'nhost',
      'https://hasura.io/jwt/claims': {
        'x-hasura-user-id': 'user-123',
        'x-hasura-default-role': 'user',
        'x-hasura-allowed-roles': '{user,admin}',
      },
    };

    const token = createJWT(payload);
    const decoded = decodeUserSession(token);

    expect(decoded.sub).toBe('user-123');
    expect(decoded.iss).toBe('nhost');
    expect(decoded.iat).toBe(1700000000 * 1000);
    expect(decoded.exp).toBe(1700003600 * 1000);

    const claims = decoded['https://hasura.io/jwt/claims'] as Record<
      string,
      unknown
    >;
    expect(claims['x-hasura-user-id']).toBe('user-123');
    expect(claims['x-hasura-default-role']).toBe('user');
    expect(claims['x-hasura-allowed-roles']).toEqual(['user', 'admin']);
  });

  it('should decode a JWT whose base64url payload contains - and _ characters (issue #3890)', () => {
    // This payload mimics a real-world scenario: a Google OAuth user with an avatar URL.
    // The avatar URL and other claim values cause the base64url-encoded payload to
    // contain `-` and `_` characters, which are valid in base64url (RFC 4648 Section 5)
    // but NOT valid in standard base64 (which uses `+` and `/` instead).
    // The browser's native atob() only supports standard base64, so passing
    // base64url directly to atob() throws InvalidCharacterError.
    const payload = {
      sub: 'f5765cb0-5b62-4fc0-b1a5-3e8e12345678',
      iat: 1700000000,
      exp: 1700003600,
      iss: 'nhost',
      'https://hasura.io/jwt/claims': {
        'x-hasura-user-id': 'f5765cb0-5b62-4fc0-b1a5-3e8e12345678',
        'x-hasura-default-role': 'user',
        'x-hasura-allowed-roles': '{user}',
        'x-hasura-avatar-url':
          'https://lh3.googleusercontent.com/a/ACg8ocLv_IKZLRBq7-xKP3BxhGvJnQzXYs96?sz=200',
      },
    };

    const token = createJWT(payload);
    const encodedPayload = token.split('.')[1] as string;

    // Precondition: verify the base64url-encoded payload actually contains
    // the problematic characters that trigger the bug.
    expect(encodedPayload.includes('-') || encodedPayload.includes('_')).toBe(
      true,
    );

    // This call should succeed but currently throws InvalidCharacterError
    // in browser environments (and in Node.js 16+ where atob is globally available)
    // because atob() does not support base64url encoding.
    const decoded = decodeUserSession(token);

    expect(decoded.sub).toBe('f5765cb0-5b62-4fc0-b1a5-3e8e12345678');
    expect(decoded.iat).toBe(1700000000 * 1000);
    expect(decoded.exp).toBe(1700003600 * 1000);

    const claims = decoded['https://hasura.io/jwt/claims'] as Record<
      string,
      unknown
    >;
    expect(claims['x-hasura-avatar-url']).toBe(
      'https://lh3.googleusercontent.com/a/ACg8ocLv_IKZLRBq7-xKP3BxhGvJnQzXYs96?sz=200',
    );
  });

  it('should throw on invalid token format', () => {
    expect(() => decodeUserSession('not-a-jwt')).toThrow(
      'Invalid access token format',
    );
    expect(() => decodeUserSession('only.two')).toThrow(
      'Invalid access token format',
    );
    expect(() => decodeUserSession('')).toThrow('Invalid access token format');
  });
});

File: packages/nhost-go/session/session_test.go
package session_test

import (
	"encoding/base64"
	"encoding/json"
	"testing"

	"github.com/nhost/nhost/packages/nhost-go/auth"
	"github.com/nhost/nhost/packages/nhost-go/session"
)

func makeToken(t *testing.T, payload map[string]any) string {
	t.Helper()

	header := base64.RawURLEncoding.EncodeToString([]byte(`{"alg":"HS256","typ":"JWT"}`))

	raw, err := json.Marshal(payload)
	if err != nil {
		t.Fatalf("marshal payload: %v", err)
	}

	body := base64.RawURLEncoding.EncodeToString(raw)

	return header + "." + body + ".sig"
}

func TestDecodeUserSessionPostgresArrayClaims(t *testing.T) {
	t.Parallel()

	token := makeToken(t, map[string]any{
		"exp": 9999999999,
		"iat": 1000,
		"sub": "user-1",
		"https://hasura.io/jwt/claims": map[string]any{
			"x-hasura-default-role":  "user",
			"x-hasura-allowed-roles": "{user,me}",
		},
	})

	decoded, err := session.DecodeUserSession(token)
	if err != nil {
		t.Fatalf("decode: %v", err)
	}

	if decoded.Sub != "user-1" || decoded.Exp != 9999999999 {
		t.Fatalf("sub/exp = %q/%d", decoded.Sub, decoded.Exp)
	}

	roles, ok := decoded.HasuraClaims["x-hasura-allowed-roles"].([]string)
	if !ok {
		t.Fatalf("allowed-roles type = %T", decoded.HasuraClaims["x-hasura-allowed-roles"])
	}

	if len(roles) != 2 || roles[0] != "user" || roles[1] != "me" {
		t.Fatalf("roles = %v, want [user me]", roles)
	}

	if decoded.HasuraClaims["x-hasura-default-role"] != "user" {
		t.Fatalf("default-role = %v", decoded.HasuraClaims["x-hasura-default-role"])
	}
}

func TestDecodeUserSessionInvalid(t *testing.T) {
	t.Parallel()

	if _, err := session.DecodeUserSession("not-a-jwt"); err == nil {
		t.Fatal("expected error for malformed token")
	}
}

func TestStorageSetGetRemove(t *testing.T) {
	t.Parallel()

	token := makeToken(t, map[string]any{"exp": 9999999999, "sub": "u"})
	store := session.NewStorage(&session.MemoryStorage{})

	if err := store.Set(auth.Session{
		AccessToken:  token,
		RefreshToken: "r",
	}); err != nil {
		t.Fatalf("set: %v", err)
	}

	got, ok := store.Get()
	if !ok || got.AccessToken != token {
		t.Fatalf("get after set failed: ok=%v", ok)
	}

	if got.DecodedToken.Sub != "u" {
		t.Fatalf("decoded token not derived on set: sub=%q", got.DecodedToken.Sub)
	}

	store.Remove()

	if _, ok := store.Get(); ok {
		t.Fatal("session present after remove")
	}
}
